File: sdks/go/pkg/beam/io/fileio/read_test.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fileio

import (
	"testing"

	"github.com/apache/beam/sdks/v2/go/pkg/beam"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/passert"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/ptest"
)

func TestReadMatches(t *testing.T) {
	tests := []struct {
		name    string
		opts    []ReadOptionFn
		input   []any
		want    []any
		wantErr bool
	}{
		{
			name: "Read matches",
			input: []any{
				FileMetadata{
					Path: "file1.txt",
					Size: 5,
				},
				FileMetadata{
					Path: "file2.txt",
					Size: 0,
				},
			},
			want: []any{
				ReadableFile{
					Metadata: FileMetadata{
						Path: "file1.txt",
						Size: 5,
					},
					Compression: compressionAuto,
				},
				ReadableFile{
					Metadata: FileMetadata{
						Path: "file2.txt",
						Size: 0,
					},
					Compression: compressionAuto,
				},
			},
		},
		{
			name: "Read matches with specified compression",
			opts: []ReadOptionFn{
				ReadGzip(),
			},
			input: []any{
				FileMetadata{
					Path: "file1",
					Size: 5,
				},
				FileMetadata{
					Path: "file2",
					Size: 0,
				},
			},
			want: []any{
				ReadableFile{
					Metadata: FileMetadata{
						Path: "file1",
						Size: 5,
					},
					Compression: compressionGzip,
				},
				ReadableFile{
					Metadata: FileMetadata{
						Path: "file2",
						Size: 0,
					},
					Compression: compressionGzip,
				},
			},
		},
		{
			name: "Read matches and skip directories",
			input: []any{
				FileMetadata{
					Path: "dir/",
					Size: 0,
				},
				FileMetadata{
					Path: "file1.txt",
					Size: 5,
				},
			},
			want: []any{
				ReadableFile{
					Metadata: FileMetadata{
						Path: "file1.txt",
						Size: 5,
					},
					Compression: compressionAuto,
				},
			},
		},
		{
			name: "Error - directories disallowed",
			opts: []ReadOptionFn{
				ReadDirectoryDisallow(),
			},
			input: []any{
				FileMetadata{
					Path: "dir/",
					Size: 0,
				},
				FileMetadata{
					Path: "file1.txt",
					Size: 5,
				},
			},
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p, s := beam.NewPipelineWithRoot()

			col := beam.Create(s, tt.input...)
			got := ReadMatches(s, col, tt.opts...)

			passert.Equals(s, got, tt.want...)
			if err := ptest.Run(p); (err != nil) != tt.wantErr {
				t.Errorf("ReadMatches() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func Test_isDirectory(t *testing.T) {
	tests := []struct {
		name string
		path string
		want bool
	}{
		{
			name: "Path to directory with forward slash directory separator",
			path: "path/to/",
			want: true,
		},
		{
			name: "Path to directory with backslash directory separator",
			path: "path\\to\\",
			want: true,
		},
		{
			name: "Path to file",
			path: "path/to/file",
			want: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := isDirectory(tt.path); got != tt.want {
				t.Errorf("isDirectory() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: runners/kafka-streams/src/test/java/org/apache/beam/runners/kafka/streams/translation/ReadTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.runners.kafka.streams.translation;

import static org.hamcrest.CoreMatchers.hasItems;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import java.util.List;
import org.apache.beam.runners.kafka.streams.KafkaStreamsTestRunner;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.CountingSource;
import org.apache.beam.sdk.io.Read;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.ParDo;
import org.junit.Test;

/**
 * End-to-end test for {@link ReadTranslator}: a primitive bounded {@code Read} over a {@link
 * CountingSource} feeds a recording ParDo executed in the in-process (EMBEDDED) Java SDK harness.
 * Verifies every source element reaches the DoFn.
 *
 * <p>{@code KafkaStreamsTestRunner.translate} forces the {@code Read.Bounded} into the deprecated
 * primitive read the runner translates, rather than the default splittable-DoFn expansion — so this
 * exercises the {@link ReadProcessor} + {@link ReadTranslator} path end to end.
 */
public class ReadTest {

  /** Records every element the harness feeds it so the test can assert the read produced them. */
  private static class RecordingFn extends DoFn<Long, Long> {
    private final SharedTestCollector<Long> collector;

    RecordingFn(SharedTestCollector<Long> collector) {
      this.collector = collector;
    }

    @ProcessElement
    public void processElement(@Element Long input, OutputReceiver<Long> out) {
      collector.record(input);
      out.output(input);
    }
  }

  @Test
  public void boundedReadEmitsEverySourceElement() {
    try (SharedTestCollector<Long> collector = SharedTestCollector.create()) {
      Pipeline pipeline = Pipeline.create(KafkaStreamsTestRunner.testOptions());
      pipeline
          .apply("read", Read.from(CountingSource.upTo(5)))
          .apply("record", ParDo.of(new RecordingFn(collector)));

      KafkaStreamsTestRunner.run(pipeline);

      List<Long> recorded = collector.recorded();
      // CountingSource.upTo(5) yields 0..4 exactly once each.
      assertThat(recorded.size(), is(5));
      assertThat(recorded, hasItems(0L, 1L, 2L, 3L, 4L));
    }
  }
}
